provider/maas/devices.go

```diff
@@ -324,3 +320,272 @@ func (env *maasEnviron) deviceInterfaceInfo2(deviceID string, nameToParentName m
 	return interfaceInfo, nil
 }
+
+type deviceCreatorParams struct {
+	Name                 string
+	Subnet               gomaasapi.Subnet // may be nil
+	PrimaryMACAddress    string
+	PrimaryNICName       string
+	DesiredInterfaceInfo []network.InterfaceInfo
+	CIDRToMAASSubnet     map[string]gomaasapi.Subnet
+	CIDRToStaticRoutes   map[string][]gomaasapi.StaticRoute
+	Machine              gomaasapi.Machine
+}
+
+func (env *maasEnviron) createAndPopulateDevice(params deviceCreatorParams) (gomaasapi.Device, error) {
+	createDeviceArgs := gomaasapi.CreateMachineDeviceArgs{
+		Hostname:      params.Name,
+		MACAddress:    params.PrimaryMACAddress,
+		Subnet:        params.Subnet, // can be nil
+		InterfaceName: params.PrimaryNICName,
+	}
+	device, err := params.Machine.CreateDevice(createDeviceArgs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	interface_set := device.InterfaceSet()
+	if len(interface_set) != 1 {
+		// Shouldn't be possible as machine.CreateDevice always
+		// returns a device with one interface.
+		names := make([]string, len(interface_set))
+		for i, iface := range interface_set {
+			names[i] = iface.Name()
+		}
+		return nil, errors.Errorf("unexpected number of interfaces "+
+			"in response from creating device: %v", names)
+	}
+	primaryNIC := interface_set[0]
+	primaryNICVLAN := primaryNIC.VLAN()
+
+	// Populate the rest of the desired interfaces on this device
+	for _, nic := range params.DesiredInterfaceInfo {
+		if nic.InterfaceName == params.PrimaryNICName {
+			// already handled in CreateDevice
+			continue
+		}
+		// We have to register an extra interface for this container
+		// (aka 'device'), and then link that device to the desired
+		// subnet so that it can acquire an IP address from MAAS.
+		createArgs := gomaasapi.CreateInterfaceArgs{
+			Name:       nic.InterfaceName,
+			MTU:        nic.MTU,
+			MACAddress: nic.MACAddress,
+		}
+
+		subnet, knownSubnet := params.CIDRToMAASSubnet[nic.CIDR]
+		if !knownSubnet {
+			logger.Warningf("NIC %v has no subnet - setting to manual and using 'primaryNIC' VLAN %d", nic.InterfaceName, primaryNICVLAN.ID())
+			createArgs.VLAN = primaryNICVLAN
+		} else {
+			createArgs.VLAN = subnet.VLAN()
+			logger.Infof("linking NIC %v to subnet %v - using static IP", nic.InterfaceName, subnet.CIDR())
+		}
+
+		createdNIC, err := device.CreateInterface(createArgs)
+		if err != nil {
+			return nil, errors.Annotate(err, "creating device interface")
+		}
+		logger.Debugf("created device interface: %+v", createdNIC)
+
+		if !knownSubnet {
+			// If we didn't request an explicit subnet, then we
+			// don't need to link the device to that subnet
+			continue
+		}
+
+		linkArgs := gomaasapi.LinkSubnetArgs{
+			Mode:   gomaasapi.LinkModeStatic,
+			Subnet: subnet,
+		}
+
+		if err := createdNIC.LinkSubnet(linkArgs); err != nil {
+			logger.Warningf("linking NIC %v to subnet %v failed: %v", nic.InterfaceName, subnet.CIDR(), err)
```

Comment:
I honestly don't know why this error isn't propagated. I just moved this code from the previous location.

Its possible it didn't want to disable an entire container because it failed to get an IP address on one of its devices.
I don't, offhand, have a good feeling for how this might fail in the wild. Its possible it is simply 'this function can return an err, so we should do something with it'.
I'm hesitant to change it today, as it has been like this for a while (since 2.0.3).